xarray/core/accessors.py

```diff
@@ -0,0 +1,103 @@
+        Array-like container of datetime-like values
+    name : str
+        Name of datetime field to access
+
+    """
+    series = pd.Series(array.ravel())
+    field_values = getattr(series.dt, name).values
+    return field_values.reshape(array.shape)
+
+_date_field_ops_and_descrs = [
+    ('year', "The year of the datetime"),
+    ('month', "The month as January=1, December=12"),
+    ('day', "The days of the datetime"),
+    ('hour', "The hours of the datetime"),
+    ('minute', "The minutes of the datetime"),
+    ('second', "The seconds of the datetime"),
+    ('microsecond', "The microseconds of the datetime"),
+    ('nanosecond', "The nanoseconds of the datetime"),
+    ('weekofyear',  "The week ordinal of the year"),
+    ('week', "The week ordinal of the year"),
+    ('weekday', "The day of the week with Monday=0, Sunday=6"),
+    ('weekday_name',
+     "The name of day in a week (ex: Friday)"),
+    ('dayofweek', "The day of the week with Monday=0, Sunday=6"),
+    ('dayofyear',  "The ordinal day of the year"),
+    ('quarter', "The quarter of the date"),
+    ('days_in_month', "The number of days in the month"),
+    ('daysinmonth', "The number of days in the month"),
+]
+
+def inject_date_field_accessors(cls):
+    for op, descr in _date_field_ops_and_descrs:
+        prop = cls._date_field_accessor(op, docstring=descr)
+        setattr(cls, op, prop)
+
+
+@register_dataarray_accessor('dt')
+class DatetimeAccessor(object):
+    """Access datetime fields for DataArrays with datetime-like dtypes.
+
+     Similar to pandas, fields can be accessed through the `.dt` attribute
+     for applicable DataArrays:
+
+        >>> ds = xarray.Dataset({'time': pd.date_range(start='2000/01/01',
+        ...                                            freq='D', periods=100)})
+        >>> ds.time.dt
+        <xarray.core.accessors.DatetimeAccessor at 0x10c369f60>
+        >>> ds.time.dt.dayofyear[:5]
+        <xarray.DataArray 'dayofyear' (time: 5)>
+        array([1, 2, 3, 4, 5], dtype=int32)
+        Coordinates:
+          * time     (time) datetime64[ns] 2000-01-01 2000-01-02 2000-01-03 ...
+
+     All of the pandas fields are accessible here. Note that these fields are not
+     calendar-aware; if your datetimes are encoded with a non-Gregorian calendar
+     (e.g. a 360-day calendar) using netcdftime, then some fields like
+     `dayofyear` may not be accurate.
+
+     """
+    def __init__(self, xarray_obj):
+        if not is_datetime_like(xarray_obj.dtype):
+            raise TypeError("'dt' accessor only available for "
+                            "DataArray with datetime64 or timedelta64 dtype")
+        self._obj = xarray_obj
+        self._dt = None
+
+    @property
+    def dt(self):
+        """Attribute to cache a view of the underlying datetime-like
+        array for passing to pandas.tslib for date_field operations
+        """
+        if self._dt is None:
+            self._dt = self._obj.values
```

Comment:
Gotcha. A few more lines of code to add to the example before:

``` python
times_arr = np.random.choice(times, size=(10, 10, nt))

dstest = datest.to_dataset()
dstest['all_times'] = (('lon', 'lat', 'time'), times_arr)

dstest = dstest.chunk({'lon': 5, 'lat': 5, 'time': 5000})
```

So now we're definitely holding a dask array:

``` python
>>> dstest.all_times
<xarray.DataArray 'all_times' (lon: 10, lat: 10, time: 100000)>
dask.array<xarray-..., shape=(10, 10, 100000), dtype=datetime64[ns], chunksize=(5, 5, 5000)>
Coordinates:
  * lon      (lon) float64 0.0 1.222 2.444 3.667 4.889 6.111 7.333 8.556 ...
  * lat      (lat) float64 0.0 2.222 4.444 6.667 8.889 11.11 13.33 15.56 ...
  * time     (time) datetime64[ns] 2000-01-01 2000-01-02 2000-01-03 ...
    time2    (time) datetime64[ns] 2000-01-01 2000-01-02 2000-01-03 ...
```

but our datetime accessors still work just fine:

``` python
>>> dstest.all_times.dt.month
<xarray.DataArray 'month' (lon: 10, lat: 10, time: 100000)>
array([[[ 3,  5, ..., 11,  5],
        [ 2, 12, ...,  8,  4],
        ..., 
        [ 9, 11, ...,  6, 11],
        [ 6,  1, ..., 12,  8]],
...
```

[Here](https://github.com/darothen/xarray/blob/8c38bf9c77f97c3e55ff175b27e4e9f659a2ba3f/xarray/core/accessors.py#L55) I cache the `values` of the DataArray passed in, and that's what the accessor checks to figure out the requested field. Are those always eagerly evaluated? If so, then it would make sense why everything works out okay - we've stored an array in memory, and we're converting that to a Pandas Series. 

Things also work fine if I create a chunked dask array from `times_arr` above and store it directly in the Dataset.